designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,544 @@
+
+```js
+const standardConfig = require("eslint-config-standard");
+const personalConfig = require("@personal/eslint-config");
+const { Config } = require("@eslint/config");
+
+const myConfig = Config.create(standardConfig, {
+    rules: {
+        semi: "warn"
+    }
+});
+
+Config.assign(myConfig, personalConfig);
+
+exports.config = myConfig;
+```
+
+This code example merges `personalConfig` into `myConfig` using the same functionality that ESLint currently uses to merge configs.
+
+#### The `RuleLoader` Class
+
+The purpose of the `RuleLoader` class is aid in loading rules to replace `--rulesdir` functionality and has this form:
+
+```js
+class RuleLoader {
+
+    // create a new instance
+    constructor() {}
+
+    // mimic `--rulesdir` loading
+    loadRulesFromDirectory(directory) {}
+}
+```
+
+In order to recreate the functionality of `--rulesdir`, a user would need to create a new entry in `ruledefs` and then specify the rules from a directory, such as:
+
+```js
+const { Config, RuleLoader } = require("@eslint/config");
+const ruleLoader = new RuleLoader();
+
+exports.config = new Config({
+    ruledefs: {
+        custom: ruleLoader.loadFromDirectory("./custom-rules")
+    },
+    rules: {
+        "custom/my-rule": "error"
+    }
+});
+```
+
+The `RuleLoader#loadFromDirectory()` method returns an object where the keys are the rule IDs (based on the filenames found in the directory) and the values are the rule objects. Unlike today, rules loaded from a local directory must have a namespace just like plugin rules (`custom` in this example).
+
+### Advanced Configs
+
+Some users may need information from ESLint to determine the correct configuration to use. To allow for that, `exports.config` may also be a function that returns an object, such as:
+
+```js
+exports.config = (context) => {
+
+    // do something
+
+    return {
+        rules: {
+            "semi": ["error", "always"]
+        }
+    };
+
+};
+```
+
+The `context` object has the following members:
+
+* `core` - information about the ESLint core that is using the config
+    * `version` - the version of ESLint being used
+    * `hasRule(ruleId)` - determine if the given rule is in the core
+    * `getConfig(configId)` - returns one of the predefined ESLint configs 
+* `cwd` - the current working directory for ESLint (might be different than `process.cwd()`, see https://github.com/eslint/eslint/issues/11218)
+
+This information allows users to make logical decisions about how the config should be constructed.
+
+#### Checking for Rule Existence
+
+One of the problems with shareable configs today is when a new rule is added to the ESLint core, shareable configs using that rule are not valid for older versions of ESLint (because ESLint validates that configured rules are present). With advanced configs, a shareable config could detect if a new rule is present before deciding to include it, for example:
```

Comment:
What about a new rule in the plugin? Would context automatically pick up all of the rules from plugins declared in `ruledefs`?

Reply:
`context` is an argument to the config function, so `ruledefs` wouldn't exist at that point since it's generated by the config. Presumably, a config that has a plugin as a dependency would already be able to figure out which rules it has, since it's providing the rules at runtime.

Reply:
Correct.